### PR TITLE
Integration test cases for Conviva Integration with Bitmovin V2 player SDK

### PR DESCRIPTION
**Pull request by luckygoyal-bitmovin**
- Added Android instrumentation test cases for validating Conviva integration module
- The test cases use Mockk framework and can be run only on devices with Android API level >= 28
- The test cases will allow to check and validate behaviour changes when migration this integration to V3 player SDK

### CODE REVIEW

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    override fun onDestroy() {
+        bitmovinPlayerView!!.onDestroy()
+        super.onDestroy()
+    }
+
+    companion object {
+        const val AUTOPLAY_KEY = "autoplay"
+        const val VMAP_KEY = "vmapTag"
+    }
+}
```
Comment (stonko1994) [inline ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt:97]: Please add a newline at EOF

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java [hunk 2]
```diff
@@ -88,6 +88,14 @@ public ConvivaAnalytics(BitmovinPlayer player,
...
                             Context context,
                             ConvivaConfiguration config) {
+        this(player, customerKey, context, new ConvivaConfiguration(), null);
+    }
+
+    public ConvivaAnalytics(BitmovinPlayer player,
+                            String customerKey,
+                            Context context,
+                            ConvivaConfiguration config,
+                            Client client) {
```
Comment (stonko1994) [inline conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:98]: I see why you need this however I'm not very happy with this being a public initializer 🤔 

I'm not into Java but can we make this an internal initializer? (not exposed through the public API)

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 3]
```diff
@@ -0,0 +1,1570 @@
...
+        tearDownMocks()
+    }
+
+    fun mockClientSettingsObject() : ClientSettings {
+        var clientSettings = ClientSettings("customerKey")
+        if (convivaConfig?.getGatewayUrl() != null) {
+            clientSettings?.gatewayUrl = convivaConfig?.getGatewayUrl()
+        }
+        mockkObject(clientSettings)
+        return clientSettings;
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt:78]:
```suggestion
        return clientSettings
```

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 4]
```diff
@@ -0,0 +1,1570 @@
...
+        systemSettings.allowUncaughtExceptions = false
+
+        if (convivaConfig?.isDebugLoggingEnabled() == true) {
+            systemSettings.logLevel = SystemSettings.LogLevel.DEBUG
+        }
+
+        val androidSystemFactory = SystemFactory(androidSystemInterface, systemSettings)
+        var client = Client(clientSettings, androidSystemFactory)
+        mockkObject(client)
+        return client;
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt:93]:
```suggestion
        return client
```

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 5]
```diff
@@ -0,0 +1,1570 @@
...
+
+        return metadata
+    }
+
+    fun expectedContentMetadata(player: BitmovinPlayer,
+                                source: SourceItem,
+                                streamType: ContentMetadata.StreamType,
+                                duration: Int,
+                                overrideMetadata: MetadataOverrides,
+                                overrideCustom: Boolean = false) : ContentMetadata {
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt:165]: I would prefer a code style like this:
```suggestion
    fun expectedContentMetadata(
        player: BitmovinPlayer,
        source: SourceItem,
        streamType: ContentMetadata.StreamType,
        duration: Int,
        overrideMetadata: MetadataOverrides,
        overrideCustom: Boolean = false
    ) : ContentMetadata {
```

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 6]
```diff
@@ -0,0 +1,1570 @@
...
+        } else {
+            false
+        }
+    }
+
+    @Test
+    fun test_implicitSessionStartOnManualPlay() {
+        val activityScenario = ActivityScenario.launch(MainActivity::class.java)
+        activityScenario.onActivity { activity: MainActivity ->
+            try {
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt:214]: Do we really need this in every test? 🤔 looks like tedious boilerplate code. Any chance to get rid of this? Would it be possible to move the `activity` handling into `createConvivaAnalyticsObject`?

File: ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt (hunk 1 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/java/com/bitmovin/analytics/conviva/testapp/MainActivity.kt]: Addressed some comments
- Split in to separate files
- Moved all common code in a base class

File: conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java (hunk 2 above)
Comment (luckygoyal-bitmovin) [reply conviva/src/main/java/com/bitmovin/analytics/conviva/ConvivaAnalytics.java:98]: Did not find solution to overcome this

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt (hunk 3 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: done

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt (hunk 4 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: done

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt (hunk 5 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: done

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt (hunk 6 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: Removed most of boiler plate code

File: ConvivaTestApp/src/main/res/xml/network_security_config.xml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<network-security-config>
+<debug-overrides>
+    <trust-anchors>
+        <!-- Trust user added CAs while debuggable only -->
+        <certificates src="user" />
+    </trust-anchors>
+</debug-overrides>
+</network-security-config>
```
[the remark this answers is not in the document]
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/main/res/xml/network_security_config.xml:9]: done

File: build.gradle [hunk 8]
```diff
@@ -27,4 +28,4 @@ allprojects {
 
 task clean(type: Delete) {
     delete rootProject.buildDir
-}
+}
```
[the remark this answers is not in the document]
Comment (luckygoyal-bitmovin) [reply build.gradle]: done

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 9]
```diff
@@ -0,0 +1,1570 @@
...
+        val customInternTags: MutableMap<String, String> = HashMap()
+        customInternTags["contentType"] = "Episode"
+        customInternTags["streamType"] = "CUSTOM_STREAM_TYPE"
+        customInternTags["integrationVersion"] = "CUSTOM_INTEGRATION_VERSION"
+        metadata.custom = customInternTags
+        metadata.streamType = ContentMetadata.StreamType.UNKNOWN
+        metadata.streamUrl = "CUSTOM_STREAM_URL"
+
+        return metadata
+    }
```
[the remark this answers is not in the document]
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: done

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt [hunk 10]
```diff
@@ -0,0 +1,1570 @@
...
+import java.security.SecureRandom
+import java.util.*
+
+@RunWith(AndroidJUnit4::class)
+class ConvivaAnalyticsTest {
+    var convivaConfig: ConvivaConfiguration? = null
+    var convivaAnalytics: ConvivaAnalytics? = null
+    var clientSettingsMock: ClientSettings? = null
+    var clientMock: Client? = null
+    var playerStateManagerMock: PlayerStateManager? = null
```
[the remark this answers is not in the document]
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/ConvivaAnalyticsTest.kt]: Removed forced un-wrapping where ever possible.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                clientMock?.adStart(
+                    CONVIVA_SESSION_ID,
+                    Client.AdStream.SEPARATE,
+                    Client.AdPlayer.CONTENT,
+                    Client.AdPosition.PREROLL
+                )
+            }
+        }
+
+        activityScenario.close()
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt:40]: Is there something like an `afterEach` for the used testing framework?

What could be possible is to add the `activeScenario` as an optional instance variable to the `TestBase` and call that in a `afterEach` method inside the `TestBase` file.

This should enable you to remove this line from all tests.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+        convivaConfig = ConvivaConfiguration()
+        convivaConfig.isDebugLoggingEnabled = true
+    }
+
+    @Before
+    fun tearDown() {
+        convivaAnalytics?.let { it.endSession() }
+
+        // Unmock mocked objects
+        tearDownMocks()
```
Comment (stonko1994) [inline ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt:55]: Following up on my previous comment: This looks like the perfect place to `close` the `activeScenario` 🙂

**Review by stonko1994 — COMMENTED** [review]
The tests look much better now 💪 Nice improvements

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt (hunk 11 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/AdsTrackingTests.kt]: done.

File: ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt (hunk 12 above)
Comment (luckygoyal-bitmovin) [reply ConvivaTestApp/src/androidTest/java/com/bitmovin/analytics/conviva/testapp/TestBase.kt:57]: done

**Review by stonko1994 — APPROVED** [review]

**Comment by luckygoyal-bitmovin** [comment]
Merging changes as 2 approvals received.

**Pull request merged**